crates/ibc/src/lib.rs

```diff
@@ -293,62 +368,45 @@ where
                         .map_err(StorageError::new)?;
-                        let receiver = packet_data.receiver.to_string();
-                        let addr = TAddrData::Ibc(receiver.clone());
-                        accum.decoder.insert(ibc_taddr(receiver), addr);
-                        let ibc_denom = packet_data.token.denom.to_string();
-                        let amount = packet_data
-                            .token
-                            .amount
-                            .try_into()
-                            .into_storage_result()?;
-                        accum = apply_recv_msg(
-                            storage,
-                            accum,
-                            &msg,
-                            vec![ibc_denom],
-                            amount,
-                            keys_changed,
-                        )?;
-                    } else {
-                        let packet_data =
-                            serde_json::from_slice::<NftPacketData>(
-                                &msg.packet.data,
-                            )
-                            .map_err(StorageError::new)?;
-                        let receiver = packet_data.receiver.to_string();
-                        let addr = TAddrData::Ibc(receiver.clone());
-                        accum.decoder.insert(ibc_taddr(receiver), addr);
-                        let ibc_traces = packet_data
-                            .token_ids
-                            .0
-                            .iter()
-                            .map(|token_id| {
-                                ibc_trace_for_nft(
-                                    &packet_data.class_id,
-                                    token_id,
-                                )
-                            })
-                            .collect();
-                        accum = apply_recv_msg(
-                            storage,
-                            accum,
-                            &msg,
-                            ibc_traces,
-                            Amount::from_u64(1),
-                            keys_changed,
-                        )?;
-                    }
+                    let addr = TAddrData::Ibc(recv_info.receiver.clone());
+                    accum.decoder.insert(ibc_taddr(recv_info.receiver), addr);
+                    accum = apply_recv_msg(
+                        storage,
+                        accum,
+                        &msg,
+                        recv_info.ibc_traces,
+                        recv_info.amount,
+                        keys_changed,
+                    )?;
                 }
-            }
+                MsgEnvelope::Packet(PacketMsg::Ack(msg)) => {
+                    let refund_info = recv_info_from_packet(&msg.packet, true)
+                        .map_err(StorageError::new)?;
+                    accum = apply_refund_msg(
```

Comment:
Is it ok to apply the refund message without checking if the ack is successful or not?